### PR TITLE
Convert associated abspaths

### PR DESCRIPTION
**Pull request by juzen2003**
Changes made:
- Add rules for `ASSOCIATIONS` (`associations_to_bundles`, `associations_to_previews`, `associations_to_metadata`, `associations_to_documents`)
- Add tests for those rules except `associations_to_metadata` since we need to create the index files for them
- Update `run_tests_coverages.sh` to run only pds3 tests when use_shelves_only is True because we don't have shelves for pds4 yet
- Replace all `holdings` with `cls.PDS_HOLDINGS` in `pdsfile.py`
- Create previews, documents, calibrated, metadata directories at the same level as bundles in Dropbox
    - Please review the structure of reviews, documents, calibrated, metadata directories and see if they make sense.

Known issues:
- For uranus under documents directory, it's currently empty. No ideas what files we should put here.
- For metadata directories, we need to create new index files, will have to work on the script under `validation`directory.

### CODE REVIEW

**Comment by juzen2003** [comment]
- Merge current main
- Add test cases for `associations_to_metadata` rules for `uranus_occs_earthbased` (`test_associated_abspaths`)
- Create two index files for `data/rings` and `data/global`:
    - `Pds4FileTest/pds4-holdings/metadata/uranus_occs_earthbased/uranus_occ_u0_kao_91cm/occ_rings.csv`
    - `Pds4FileTest/pds4-holdings/metadata/uranus_occs_earthbased/uranus_occ_u0_kao_91cm/occ_global.csv`

File: run_tests_coverage.sh [hunk 1]
```diff
@@ -11,8 +11,16 @@ mode_desc=( [1]="use shelves" [2]="no shelves" )
 while [ $count -le $numOfModes ]
 do
     echo "Run with ${mode_desc[$count]}"
-    coverage run -a -m pytest pdsfile/pds3file/tests/ \
-    pdsfile/pds3file/rules/*.py pdsfile/pds4file/tests/ pdsfile/pds4file/rules/*.py --mode ${mode_arr[$count]}
+    # Only run with use shelves for pds3, we don't have shelves for pds4 for now
```
Comment (pds-admin) [inline run_tests_coverage.sh:14]: This file has gotten really complicated. How about just having four separate test lines with the appropriate arguments and skipping all the arrays and loops and conditionals?

Comment (rfrenchseti) [inline run_tests_coverage.sh:14]: This file has gotten really complicated. How about just having four separate test lines with the appropriate arguments and skipping all the arrays and loops and conditionals?

File: pdsfile/pdsfile.py [hunk 2]
```diff
@@ -3785,9 +3786,11 @@ def from_abspath(cls, abspath, fix_case=False, must_exist=False,
         # Search for "holdings"
         parts_lc = [p.lower() for p in parts]
         try:
-            PDS_HOLDINGS_index = parts_lc.index(cls.PDS_HOLDINGS) # Change variable name to distinguish from PDS3
+            # Change variable name to distinguish from PDS3
+            PDS_HOLDINGS_index = parts_lc.index(cls.PDS_HOLDINGS)
         except ValueError:
-            raise ValueError('"{}" directory not found in: '.format(cls.PDS_HOLDINGS) + abspath)
+            raise ValueError('"{}" directory not found in: '.format(cls.PDS_HOLDINGS)
```
Comment (rfrenchseti) [inline pdsfile/pdsfile.py:3792]: Let's get rid of uses of `format` and switch to f-strings everywhere.

File: pdsfile/pds4file/rules/uranus_occs_earthbased.py [hunk 3]
```diff
@@ -52,131 +52,41 @@
...
+    (r'.*/(uranus_occs_earthbased/uranus_occ_u.*)/(data|browse)(.*|_[a-z]*]/.*)\.[a-z]{3}', 0,
+        [r'bundles/\1/data\3.tab',
+         r'bundles/\1/data\3.xml',
+         r'bundles/\1/data\3.txt',
+         r'bundles/\1/data\3.pdf',
+         r'bundles/\1/browse\3.pdf',
+         r'bundles/\1/browse\3.xml',
+        ]),
+    (r'documents/uranus_occs_earthbased.*', 0,
+        [r'bundles/uranus_occs_earthbased',
```
Comment (rfrenchseti) [inline pdsfile/pds4file/rules/uranus_occs_earthbased.py:65]: Why is the same thing in the list three times?

File: pdsfile/pds4file/rules/uranus_occs_earthbased.py [hunk 4]
```diff
@@ -52,131 +52,41 @@
...
-    (r'.*/(COISS_3xxx.*/COISS_3...)/(data|extras/\w+)(|/images|/maps)', 0,
-            [r'previews/\1/data/\3',
-             r'previews/\1/extras/browse/\3',
-             r'previews/\1/extras/thumbnail/\3',
-             r'previews/\1/extras/full/\3',
-            ]),
-    (r'.*/(COISS_3xxx.*/COISS_3...)/extras', 0,
-            r'previews/\1/data'),
+    (r'.*/(uranus_occs_earthbased/uranus_occ_u.*/(data|browse)(.*|_[a-z]*])/.*)\.[a-z]{3}', 0,
+        [r'previews/\1_full.png',
```
Comment (rfrenchseti) [inline pdsfile/pds4file/rules/uranus_occs_earthbased.py:73]: Can leave this but I'll just point out we don't have any preview images yet.

File: pdsfile/pds4file/rules/uranus_occs_earthbased.py [hunk 5]
```diff
@@ -52,131 +52,41 @@
...
-             r'metadata/\1/\3/\3_jupiter_summary.tab/\5',
-            ]),
-    (r'metadata/(COISS_.xxx/COISS_[12])...', 0,
-            r'metadata/\g<1>999'),
-    (r'metadata/(COISS_.xxx/COISS_[12]).../(COISS_.)..._(.*)\..*', 0,
-            [r'metadata/\g<1>999/\g<2>999_\3.tab',
-             r'metadata/\g<1>999/\g<2>999_\3.csv',
-             r'metadata/\g<1>999/\g<2>999_\3.lbl',
-            ]),
+    (r'.*/(uranus_occs_earthbased)/(uranus_occ_u.*)/(data|browse)(.*|_[a-z]*])/(rings|global)/(.*)\.[a-z]{3}', 0,
```
Comment (rfrenchseti) [inline pdsfile/pds4file/rules/uranus_occs_earthbased.py:81]: There will only be one metadata directory for the whole bundleset, not a subdirectory per bundle.

**Review by rfrenchseti — CHANGES_REQUESTED** [review]
I did a basic review of the files excluding ISS and VIMS, which I don't know anything about.

**Comment by juzen2003** [comment]
Update based on comments:
- Add test case for metadata associated abspaths to test on the pdsfile instance of the entire bundle. Commented out test cases that test on the pdsfile instance of a data file (These will work when we add index shelf files later on)
- Replace `.format` with f string interpolation in `pdsfile.py`
- Simplify `run_tests_coverage.sh`

**Review by mace-space — APPROVED** [review]
Looks good

**Review by rfrenchseti — APPROVED** [review]

**Pull request merged**